### PR TITLE
Initial implementation of `dune fmt`

### PR DESCRIPTION
**Pull request by emillon**
This is a first draft of `dune fmt` (see #940) with three main limitations:

- it is language agnostic, so it does not know about field names
- it is not able to parse comments
- it does not break long lines

The formatting rules are pretty simple:

- lists composed only of atoms, quoted strings, templates, and singletons are displayed on a single line
- other lists are displayed with a line break after each element
- an empty line is inserted between toplevel stanzas

The CLI is minimal: it can either read a file or standard input, but will always output on the standard input. In addition, it is necessary to pass `--unstable` to enable this feature, until some guarantees are given.

It is possible to test the output on all the dune files in the repository using the following command (zsh; possibly bash supports that glob syntax as well):

```zsh
for x in test/**/dune ; do
    if [ ! -x $x ] ; then
        echo "; $x" >> /tmp/output
        ./_build/default/bin/main_dune.exe fmt < $x >> /tmp/output 2>&1
    fi
done
```

I am pretty satisfied with the output:

```scheme
; test/blackbox-tests/test-cases/js_of_ocaml/lib/dune
(library
 (name x)
 (public_name x)
 (js_of_ocaml
  (flags (--pretty))
  (javascript_files runtime.js)
 )
 (c_names stubs)
 (preprocess
  (pps js_of_ocaml-ppx)
 )
)

; test/blackbox-tests/test-cases/loop/dune
(rule
 (copy %{read:x} a)
)

(rule
 (copy %{read:y} b)
)

(rule
 (progn
  (run true)
  (with-stdout-to
   x
   (echo b)
  )
 )
)
```

Indenting with only one space is pretty narrow, but this works well because it aligns with the left parens.

Let me know what you think.

### CODE REVIEW

**Comment by emillon** [comment]
NB: this triggers some errors in some of the `dune` files, I intend to fix these.

**Comment by rgrinberg** [comment]
Can we maintain the standard lisp formatting? Where trailing braces are on the same line. I don't see a reason to break with tradition here.

**Comment by diml** [comment]
I like the output, but I agree with @rgrinberg about trailing parentheses. If we don't group them they occupy a lot of lines.

It would be nice if we could also preserve the original formatting of strings. Could you also add a mode where it just reformat all the dune files in the source tree in-place?

**Comment by emillon** [comment]
I don't have a lot of lisp experience, so I tried to keep this as simple as possible for now. Is something like this more idiomatic?

```scheme
(rule
 (progn
  (run true)
  (with-stdout-to
   x
   (echo b))))
```

> It would be nice if we could also preserve the original formatting of strings.

I believe that this will require an alternative parser, so I suggest that I do this in a separate PR (with comments as well).

> Could you also add a mode where it just reformat all the dune files in the source tree in-place?

Sure!

**Comment by emillon** [comment]
>  I don't see a reason to break with tradition here.

We can probably make two (related) arguments: it diffs poorly, and it's difficult to edit the end of a s-expression, for example to add a field to `(library)` or the like.

**Comment by diml** [comment]
Personally, i don't mind about breaking the tradition, all that matters is that it looks good. It's true that it might be annoying if it's hard to add a field at the end.

Do you know how we can integrate this with emacs? so that for example the file is automatically reformatted when saving the file to disk. I'd like to try to see how it feels.

**Comment by rgrinberg** [comment]
> We can probably make two (related) arguments: it diffs poorly,

What's the issue with diffing? It works quite well for me. Even better is the structural diff that is available in Emacs.

> and it's difficult to edit the end of a s-expression, for example to add a field to (library) or the like.

With what editor? In emacs this is 1 key combination. In vim it shouldn't be very hard either. Just go to the last field, hit `%i<return>` and write.

Here's another point for the classical formatting: almost everyone uses it. The vast majority of jbuild files use it. You can check it yourself with `cat $(find . -name jbuild) | less` if you don't believe me. That signals preference towards the classical style.

File: src/dune_fmt.ml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  let fname, contents =
+    match path_opt with
+    | Some path ->
+      Io.with_file_in path ~f:(fun ic ->
+        let contents = Io.read_all ic in
+        (Path.to_string path, contents)
+      )
+    | None ->
+      let lines = Io.input_lines stdin in
+      let contents = String.concat ~sep:"" lines in
```
Comment (diml) [inline src/dune_fmt.ml]: `sep` should be `"\n"`, otherwise 2 lines that contain a single atom would end up concatenated as a single one.

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by emillon** [comment]
I don't want this PR to turn into a debate on what should be the definitive syntax. The point is to start from something easy to implement and then we can improve based on feedback.

> Even better is the structural diff that is available in Emacs.

Emacs is very powerful to edit s-expressions, but not all dune users use it.

IMO having `)` characters more visible will make dune easier to use for people less used to s-expressions.

> Here's another point for the classical formatting: almost everyone uses it. The vast majority of jbuild files use it. You can check it yourself with cat $(find . -name jbuild) | less if you don't believe me. That signals preference towards the classical style.

I believe you, but this is a bit biased. For example, when contributing my first `dune` files I checked in this repo to see how this was already done here, and adjusted. But for fresh projects, I used the the more vertical style [like this one](https://github.com/emillon/ocaml-noise/blob/079e8aae9a82a76883dcf6bd0460b31d917d4bb1/test/vectors/dune) (see also [this one](https://github.com/cryptosense/key-parsers/blob/c44b8366c99f14685e0b8cb0265cdf23381460a3/lib/dune)).

> Do you know how we can integrate this with emacs? so that for example the file is automatically reformatted when saving the file to disk. I'd like to try to see how it feels.

I'm not familiar enough with elisp to provide a working example, but something like this should work:

```elisp
; warning: untested
(defun dunefmt-before-save ()
  (interactive)
  (shell-command-on-region
   (point-min) (point-max)
   "dune fmt --unstable"))

(add-hook 'before-save-hook #'dunefmt-before-save))
```

**Comment by diml** [comment]
I agree, let's not discuss this eagerly. We need to try this to see what's best in practice, so merging this as an unstable feature that we can play with seems good.

**Review by ghost — APPROVED** [review]

**Comment by rgrinberg** [comment]
Okay, we can leave the syntax aside. I still think its dubious to include the formatting tool if we can't even use it on our dune files.

Btw, do we really need this `--unstable` flag? I'd like to see some motivation for it as I think it's just bad UI to have a command that only works with a flag. We might as well rename the command to `$ dune unstable-fmt`.

**Comment by emillon** [comment]
> Btw, do we really need this `--unstable` flag?

The idea was to just remove that flag at the end rather than rename the command, but it's the same. I followed your suggestion and renamed the command.

**Review by rgrinberg — APPROVED** [review]

**Comment by emillon** [comment]
Thanks for the feedback! Can you open a separate issue for this?

**Pull request merged**